docs/design/HotReloadDoc.md

```diff
@@ -0,0 +1,49 @@
+
+One consideration is that in order to handle hot reloading of the sections outside of the runtime section we may need to walk a tree of config files. Since with recent changes we can now reference another configuration file, which can itself reference another configuration file, to be certain that we are hot reloading when any change is made to a configuration file, we will need to walk that tree and monitor all of those files. However, this will only be needed for the sections beyond the runtime section, because the root in this tree will always have the runtime data.
+
+### Hot Reloading Trigger:
+
+Define a trigger mechanism to initiate the hot reloading process when changes are detected in the config file. This trigger function will then call into the `RuntimeConfigProvider` initiating the update of the `RuntimeConfig` by reloading the new config and creating a new object model. This trigger function lives in the `ConfigFileWatcher` and during construction we register this function with the underlying `FileSystemWatcher.Changed` field. This function then calls into the `RuntimeConfigProvider` which has a hot reload function that will then use the config loader to create the new object model that we save as the `RuntimeConfig` in the `RuntimeConfigProvider`.
+
+### Config Object Update:
+
+When the hot reloading is triggered, create a new `RuntimeConfig` record using the same deserialization process as on startup. We call into this through the same function provided by the config loader that we use on startup. We use the same config file path in the same file system again, as provided by the config loader.
+
+By using the regular deserialization process we should make sure that we correctly walk through the tree of config files and correctly deserialize from all of them to create the right object model, but we need to make certain that this is the case during implementation.
+
+### Dependency Updating:
+
+Implement a mechanism to update services that are already registered during startup and depend on configuration settings that have changed. Authorization in particular is only configured on startup. In order to reset those authorization settings when we change the config file we need a way to have those particular settings reset and update. This is a detail that remains in design.
+
+# Potential Classes still in consideration
+
+## IConfigurableService Interface:
+
+### Responsibilities:
+Interface to be implemented by services that depend on configuration settings.
```

Comment:
I don't agree with this approach.. instead, i think we could consider:
-Ioptionsmonitor to detect onchange event to then modify class internals
-OR ensure that when RuntimeConfigProvider is injected in class constructors, it's internal runtimeconfig is not set in the class. Instead set the runtimeconfig provider so that any usage of the runtime config can be fed from runtimeconfig provider (runtimeconfigprovider abstracts away the need to know about any hot loading config). This points us in the right direction but still requires more thought because our runtimeconfigprovider/runtimeconfig relationship is a service locator antipattern.

Reply:
> I don't agree with this approach.. instead, i think we could consider: -Ioptionsmonitor to detect onchange event to then modify class internals
> 
> -OR ensure that when RuntimeConfigProvider is injected in class constructors, it's internal runtimeconfig is not set in the class.
> 
> * Instead set the runtimeconfig provider so that any usage of the runtime config can be fed from runtimeconfig provider (runtimeconfigprovider abstracts away the need to know about any hot loading config).
> * This points us in the right direction but still requires more thought because our runtimeconfigprovider/runtimeconfig relationship is a service locator antipattern.


After our discussion yesterday it seems that we might need multiple approaches to make sure that this work as we want with all of our services.

for example updating the jwt options seems to lend itself to something like `IOptionsMonitor` approach, but then the `OpenApiDocumentor` could be modified to use the `RuntimeConfigProvider` to get a `RuntimeConfig` rather than storing one internally to make sure it correctly reflects the hot reloads.